### DISCUSSION TITLE
Implicitly scoped lock statement

### DISCUSSION THREAD

**Discussion by skarllot** [dotnet/csharplang#2451] [Language Ideas] [answered]
## Motivation:
Allow `lock` keyword to be used to lock a variable that will be unlocked at the end of the current block. This avoids creating a new block with an additional level of indentation:

### Before:
```csharp
public void Foo() {
    lock (numberGate) {
        number += 5;
        if (number > 1000) {
            number = 0;
        }
    }
}
```

### After:
```csharp
public void Foo() {
    lock numberGate;
    number += 5;
    if (number > 1000) {
        number = 0;
    }
}
```

**Comment by YairHalberstadt** [comment]
It is absolutely vital that locks are as short as possible. As such this might be considered a bad idea.

**Comment by marksmeltzer** [comment]
Same rationale as was used for simplifying `using`, and I love that simplification.

**Comment by theunrepentantgeek** [comment, marked as answer]
@marksmeltzer wrote:

> Same rationale as was used for simplifying using, and I love that simplification.

The simplification for **using** makes sense because the vast majority of using statements finish at the end of the enclosing block:

```
{
    // code

    using (var thing = ...) 
    {
        // more code
    }
}
```

However, because of the need for **lock** statements to enclose as little as possible, my experience is that those blocks typically finish well before the enclosing block:

```
{
    // code

    lock (_thing) 
    {
        // more code
    }

    // yet more code
}
```

Creating a variation of **lock** that encourages longer lock retention would be a bad idea.

I'd also offer the observation that **using** statements are extremely common - so the new **using** declarations will benefit a lot of developers. By contrast, **lock** statements are relatively rare, so any benefit would accrue to a significantly smaller group.

**Comment by PathogenDavid** [comment]
> I'd also offer the observation that **using** statements are extremely common - so the new **using** declarations will benefit a lot of developers.

It's also more common to end up with many using statements in the same method, so using statements can keep the number of blocks and levels of indentation required under control in a method when omitting the blocks isn't possible.

**Comment by HaloFour** [comment]
Another difference between `lock` and `using` is that the latter is tied to the declaration of a variable, and the resource is disposed at the point that the variable goes out of scope.  You can't use a `using` declaration on an existing variable.  `lock` must be used on an existing variable, but the duration of the lock certainly wouldn't be associated with the scope of that variable.

**Comment by SilentSin** [comment]
> You can't use a using declaration on an existing variable.

Yes you can.

**Comment by HaloFour** [comment]
@SilentSin 

I'm referring to the new feature slated for C# 8.0:

```cs
void M() {
    using var foo = new Foo();
    // do stuff here
} // foo automatically disposed here
```

Sorry, should have clarified.

**Comment by skarllot** [comment]
@theunrepentantgeek for me when using locks, is common that the scope of locks are the method/property itself.

Another side effect is to provide a feature closer to Java synchronized methods.

**Comment by HaloFour** [comment]
@skarllot 

> Another side effect is to provide a feature closer to Java synchronized methods.  

This already exists in .NET:

```cs
[MethodImpl(MethodImplOptions.Synchronized)]
public void Method()
{
    MethodImpl();
}
```

C# doesn't expose it at the method level because it behaves identically to how `synchronized` is implemented in Java and even at that time it was known to be a bad idea.  Sadly it wasn't yet understood that locking arbitrary `object` instead of a specific lock primitive was also a bad idea.

This is really a sidebar as I assume your proposal would keep to the concept of locking on some arbitrary object and not trying to replicate Java `synchronized` methods by implicitly locking `this` or the `Type`.  When you are using `synchronized` on an arbitrary object in Java you are still required to supply a block.  It's pretty much identical to C# in that regard.

**Comment by ghord** [comment]
Please consider **fixed** for the same feature

**Comment by DavidArno** [comment]
```cs
public void Foo() 
{
    lock (numberGate) restOfFoo();

    void restOfFoo()
    {
        number += 5;
        if (number > 1000) number = 0;
    }
}
```

🤡

**Comment by DevAM-Tools** [comment]
> It is absolutely vital that locks are as short as possible. As such this might be considered a bad idea.

and 

> Creating a variation of lock that encourages longer lock retention would be a bad idea.

I think that's actually a descision of the developer and has nothing to do with the syntax. The argument misses the real motivation behind this proposal. It's about making code more readable and easy to write.

One circumstance that may complicate the implementation of the suggestion is the fact that `lock(_lockObject);` there is already a valid systax today. However it is pretty useless.

**Reply by HaloFour** [reply]
> It's about making code more readable and easy to write.

Reabability is more about clarity than verbosity.  Having a clearly defined scope for when a lock is acquired lends to readability.  If the language makes it easier to take a lock without a scope it would encourage developers to acquire longer-lived locks.

**Reply by colejohnson66** [reply]
@HaloFour To be fair, that same argument could be made against `using` statements. Allowing brace-less `using`s encourages developers to not release resources when they should.

**Reply by HaloFour** [reply]
Yes it could, but that wasn't considered as problematic as long as the resource is released.  The impact to holding a lock longer than you should can be much greater.

**Reply by theunrepentantgeek** [reply]
From my work in Go, I can attest this is very true. The closest Go has to a direct equivilen of a **lock** statement is the ability to use **defer** to free up a `sync.Mutex` at the end of the function. I've found and fixed several deadlock issues caused by this - holding the lock too long is problematic.

The fix is usually to extract the critical section into a different function, but doing that can be messy.

**Reply by theunrepentantgeek** [reply]
> the fact that lock(_lockObject); there is already a valid systax today. However it is pretty useless.

It is pretty useless _today_.

But if #7104 goes ahead, different implementations will be free to implement the pattern with different semantics. You could, for example, easily write an implementation that implements rate throttling - by blocking callers to prevent flooding another subsystem.

(And #7104 has a proposal champion, so it's far more likely to go ahead than most ideas.)

**Reply by 333fred** [reply]
> We generally accept the changes for special casing how System.Threading.Lock interacts with the lock keyword, but we will not adopt the full pattern at this time.

At this point, we're not doing a full pattern. Just special-casing `System.Threading.Lock`.

**Reply by colejohnson66** [reply]
@333fred That's fine with me, but is there a reason the team abandoned/pushed off the more generalized `ILock` (or whatever interface it was) pattern?

**Reply by 333fred** [reply]
We didn't see the use case. The notes linked from #7104 have the full context.

**Reply by jaredpar** [reply]
> That's fine with me, but is there a reason the team abandoned/pushed off the more generalized ILock 

In the case of `Lock` the holder type is a `ref struct`. When designing this feature we hadn't committed to allowing `ref struct` in generics hence we couldn't define a `ILock<T>` that works with the most important type.

It's possible we could revisit this in the future but right now we don't have a motivating example.

**Reply by Yevhen-Myroshnychenko** [reply]
it would be nice to have option to declare local variable, class field or may be property, using syntax like:

```
lock var varName = value
```
```
class SomeClass 
{
    lock Type somField = value;
}

```
were ` lockObject = Lock()` for this use cases will be automatically created implicitly and every invocation of the `somField` will be automatically wrapped into` lock(lockObject) { somField = ...}` or `lock(lockObject) { someVarible = somField}`

**Reply by theunrepentantgeek** [reply]
Other languages have tried this and it has seldom (maybe never) turned out well. (I think Java's **synchronized** keywordk is one example, but it's been years so I might be mistaken).

The problem is that developers assume the **lock** means it's threadsafe, so they write things like this:

``` csharp
// cache is a field within a class
private lock ICache cache 

// elsewhere
if cache == nil {
    cache = NewCache()
}
```

The problem is, the lock only protects read and write, so the code actually means this:

``` csharp
// cache is a field within a class
private ICache cache 

// elsewhere
ICache temp
lock {
    temp = cache
}

if temp == nil {
    lock {
        cache = NewCache()
    }
}
```

Which has all the race condition possibilities of unprotected code. 

Worse, because taking and releasing all those locks will affect performance, adding **lock** might change runtime conditions enough to move a race condition and thus appear like a fix.

The best (correct?) scope for a **lock** is always small, but seldom as small as a single read/write.
